vowpalwabbit/accumulate.cc

```diff
@@ -39,20 +42,22 @@ float accumulate_scalar(vw& all, float local_sum)
   return temp;
 }
 
-void accumulate_avg(vw& all, regressor& reg, size_t o)
+void accumulate_avg(vw& all, weight_vector& weights, size_t o)
 { uint32_t length = 1 << all.num_bits; //This is size of gradient
-  size_t stride = 1 << all.reg.stride_shift;
-  float* local_grad = new float[length];
-  weight* weights = reg.weight_vector;
   float numnodes = (float)all.all_reduce->total;
-
-  for(uint32_t i = 0; i < length; i++)
-    local_grad[i] = weights[stride*i+o];
-
-  all_reduce<float, add_float>(all, local_grad, length);
-  for (uint32_t i = 0; i < length; i++)
-    weights[stride*i+o] = local_grad[i]/numnodes;
-  delete[] local_grad;
+  weight_vector local_grad(length);
+
+  weight_vector::iterator i = weights.begin(o);
+  weight_vector::iterator j = local_grad.begin();
+  for (; i != weights.end(); ++i, ++j)
+	  *j = *i;
+
+  all_reduce<float, add_float>(all, local_grad.first(), length); //need to be modified to not use first()
+ 
+  i = weights.begin(o);
+  j = local_grad.begin();
+  for (; i != weights.end(); ++i, ++j)
+	  *i = *j/numnodes;
 }
```

Comment:
This can be written as

`std::transform(local_grad.begin(), local_grad.end(), weights.begin(o), [numnodes](float weight) { return weight/numnodes; });`